db/dcp_sharded_upgrade_test.go

```diff
@@ -0,0 +1,283 @@
+	bucketUUID, err := tb.UUID()
+	require.NoError(t, err, "get bucket UUID")
+
+	numVBuckets, err := tb.GetMaxVbno()
+	require.NoError(t, err)
+
+	const numPartitions = 4
+
+	require.NoError(t, tb.SetRaw(base.SyncPrefix+"cfgindexDefs", 0, nil, []byte(fmt.Sprintf(indexDefs, tb.GetName(), bucketUUID))))
+	require.NoError(t, tb.SetRaw(base.SyncPrefix+"cfgnodeDefs-known", 0, nil, []byte(nodeDefs)))
+	require.NoError(t, tb.SetRaw(base.SyncPrefix+"cfgnodeDefs-wanted", 0, nil, []byte(nodeDefs)))
+	planPIndexesJSON := preparePlanPIndexesJSON(t, tb, numVBuckets, numPartitions)
+	require.NoError(t, tb.SetRaw(base.SyncPrefix+"cfgplanPIndexes", 0, nil, []byte(planPIndexesJSON)))
+
+	// Write a doc before starting the dbContext to check that import works
+	const (
+		testDoc1 = "testDoc1"
+		testDoc2 = "testDoc2"
+	)
+	require.NoError(t, tb.SetRaw(testDoc1, 0, nil, []byte(`{}`)))
+
+	t.Log("TEST: setup complete")
+
+	// Start this retry loop asynchronously because the change may happen before NewDatabaseContext returns
+	db, err := NewDatabaseContext(tb.GetName(), tb, true, DatabaseContextOptions{
+		GroupID:                    "",
+		EnableXattr:                true,
+		UseViews:                   base.TestsDisableGSI(),
+		skipStartHeartbeatChecking: true,
+		ImportOptions: ImportOptions{
+			ImportPartitions: numPartitions,
+		},
+	})
+	require.NoError(t, err, "NewDatabaseContext")
+	defer db.Close()
+
+	//pIndexes, _, err := cbgt.CfgGetPlanPIndexes(db.CfgSG)
+	//require.NoError(t, err)
+	// TODO: sometimes cbgt adds and immediately removes a PIndex causing this to fail
+	//assert.Len(t, pIndexes.PlanPIndexes, numPartitions, "number of planPIndexes should match partitions after DbContext start")
+
+	require.NoError(t, db.WaitForPendingChanges(base.TestCtx(t)))
+	doc, err := db.GetDocument(base.TestCtx(t), testDoc1, DocUnmarshalAll)
+	require.NoError(t, err, "GetDocument 1")
+	require.NotNil(t, doc, "GetDocument 1")
+
+	assert.True(t, db.ImportListener.cbgtContext.LithiumCompat.IsTrue(), "LithiumCompat")
+
+	require.NoError(t, db.Heartbeater.StartCheckingHeartbeats())
+
+	// Now wait for it to pick up that the Lithium "node" isn't actually heartbeating
+	err, _ = base.RetryLoop("wait for LithiumCompat to become false", func() (shouldRetry bool, err error, value interface{}) {
+		if db == nil {
+			return true, nil, nil
+		}
+		return db.ImportListener.cbgtContext.LithiumCompat.IsTrue(), nil, nil
+		// will take around ~10 seconds at least
+	}, base.CreateSleeperFunc(200, 10))
+	assert.NoError(t, err, "LithiumCompat never became false")
+
+	//pIndexes, _, err = cbgt.CfgGetPlanPIndexes(db.CfgSG)
+	//require.NoError(t, err)
+	// TODO: sometimes cbgt adds and immediately removes a PIndex causing this to fail
+	//assert.Len(t, pIndexes.PlanPIndexes, numPartitions, "number of planPIndexes should match partitions after LithiumCompat change")
+
+	// And write another doc to check that import still works
+	require.NoError(t, tb.SetRaw(testDoc2, 0, nil, []byte(`{}`)))
```

Comment:
Are we sure that this isn't just a doc that's written to a pindex that happens to be assigned to this node?

Reply:
Yes - initially all the PIndexes are assigned to a node that doesn't exist (unless the node under test generates the UUID `680cce1d31e72b69`, which has a chance of 1 in `16*(2^16)`)

Reply:
My question was more "are we sure that all PIndexes are getting reassigned to the new node"?  If only half are being assigned, and this doc happens to be in a vbucket in that half, the test would still pass.

Reply:
Oh, I see what you mean - added an extra check that all PIndexes are assigned to this node.